### PR TITLE
Migration per application and on core

### PR DESCRIPTION
**Pull request by sdrdis**
This pull request allows migration per application and on core.

Here are precise features :
- Migration per application : supports migration files on application migrations folder
- Migration on core : supports migrations files framework/migrations
- Supports same migration name on different application (migration files are on application namespace instead of `Fuel\Migrations`)
- Migration are executed when clicking on install button in the application manager (works for core / application / local)
- Detects if application / core / local is on last migration and shows the "Apply changes" button if not (it will execute the migrations)
- Simply supports updates from older version : if you update from an older version, the application manager will ask you to apply changes.

Please feel free to test on your own instances : pull, then go to your application manager. Only local migrations are not deleted here since it is on novius-os repo (so it will ask you to apply changes on local, but don't do it).

When pull request is accepted, I will merge novius-os repo. It would be quite complicated to post a pull request on this repo since I also updated local applications.

### CODE REVIEW

File: framework/classes/fuel/migrate.php [hunk 1]
```diff
@@ -45,4 +45,159 @@ public static function set_table($table)
         static::$table = $table;
     }
 
+    // change migration prefix and changed include to include_once in order to prevent duplicate classes errors
+    protected static function find_migrations($name, $type, $start = null, $end = null, $direction = 'up')
```
Comment (sdrdis) [inline framework/classes/fuel/migrate.php:49]: Could do a pull request on fuel for include_once

**Comment by sdrdis** [comment]
The idea is to make a pull request on find_migrations on fuel and after call parent::find_migrations.

Also i am open about static::$prefix. It is a legacy from fuel, but prefix on novius os should be unique for each application / core.

File: framework/classes/fuel/migrate.php [hunk 2]
```diff
@@ -45,4 +45,159 @@ public static function set_table($table)
...
+            static::$prefix = 'Nos\\Migrations\\';
+        } else if ($type == 'module') {
+            $namespaces = \Nos\Config_Data::get('app_namespaces', null);
+            static::$prefix = $namespaces[$name].'\\Migrations\\';
+        } else {
+            static::$prefix = 'Fuel\\Migrations\\';
+        }
+    }
+
+    public static function is_last_version($name, $type)
```
Comment (felixgilles) [inline framework/classes/fuel/migrate.php]: If not a native function of FuelPHP, use camelCase

File: framework/classes/fuel/migrate.php [hunk 3]
```diff
@@ -45,4 +45,159 @@ public static function set_table($table)
...
+        }
+
+        static::$connection === null or \DBUtil::set_connection(null);
+
+        empty($done) or logger(Fuel::L_INFO, 'Migrated to '.$ver.' successfully.');
+
+        return $done;
+    }
+
+    protected static function generate_prefix($name, $type)
```
Comment (felixgilles) [inline framework/classes/fuel/migrate.php]: If not a native function of FuelPHP, use camelCase

File: framework/classes/migration.php [hunk 4]
```diff
@@ -12,22 +12,29 @@
...
-        if (file_exists($sql_file)) {
-            $queries = explode(';', file_get_contents($sql_file)); // @todo: might not work everywhere (; in values)
-            foreach ($queries as $query) {
-                if (trim($query) != '') {
-                    // @todo: might not work for comments
-                    \DB::query($query)->execute();
-                }
+    }
+
+    public static function execute_sql_file($sql_file)
```
Comment (felixgilles) [inline framework/classes/migration.php]: If not a native function of FuelPHP, use camelCase

**Comment by sdrdis** [comment]
corrected to camelCase

**Pull request merged**